source/agora/consensus/EnrollmentManager.d

```diff
@@ -113,6 +113,9 @@ public class EnrollmentManager
     /// Ditto
     private PreImageCycle cycle;
 
+    /// Index of pre-image cycle restored from the chain
+    private uint restored_cycle_index;
```

Comment:
No. As I said countless times, we shouldn't have a special path for restoring data. That will just lead to terrible issues later. Please rethink this so we don't need to overly special-case things.

Reply:
> No. As I said countless times, we shouldn't have a special path for restoring data. That will just lead to terrible issues later. Please rethink this so we don't need to overly special-case things.

I will rewrite the code in the morning as you commented. It would be good to have a pair-programming if you have time today.

Reply:
I have changed the approach in the last commit. Please tell me how it is. The commit for network test code is removed to avoid confusion and will be contained again soon.

Still need a test.